### PR TITLE
feat: Diagram agent

### PR DESCRIPTION
**Pull request by kgilpin**
Adds `@diagram` agent mode, with built-in prompts for entity-relationship, flowchart, and class map diagrams.

1. [ ] **Diagram Generation Mode**: Introduced a new `AgentMode` called `Diagram` for generating software diagrams (flowcharts, ER diagrams, UML class diagrams).
2. [ ] **DiagramAgent**: Created a new `DiagramAgent` class for handling diagram generation requests.
3. [ ] **Agent Selection for Diagrams**: Updated `AgentSelectionService` to handle `@diagram` mode requests.
4. [ ] **Mermaid Library Integration**: Added `mermaid` library dependency for diagram rendering.
5. [ ] **MermaidValidator**: Added `MermaidValidator` to validate Mermaid diagram syntax.
6. [ ] **MermaidFilter**: Implemented `MermaidFilter` to parse and validate Mermaid diagram syntax. Tokens are withheld from the LLM until a complete diagram is received, then the diagram is validated and potentially fixed.
7. [ ] **Mermaid Fixer Service**: Added `MermaidFixerService` for repairing Mermaid diagrams.
8. [ ] **Filter Interface and NopFilter**: Defined a `Filter` interface and a `NopFilter` class. These exist to support MermaidFilter.
9. [ ] **Agent Filters**: Enabled agents to return filters for processing output through the `newFilter` method.
10. [ ] **Command Warning Enhancements**: Added color-coded warnings for questions and classifications in `ExplainCommand`.
11. [ ] **Explain Agent Updates**: Modified the `ExplainAgent` to include diagram prompt checking and added capability to generate diagrams.
12. [ ] **Classification Service Update**: Modified `ClassificationService` to handle the new classification label, and distinguish between modes and scopes.
13. [ ] **Context Label Update**: Updated `ContextV2.ContextLabelName` and `ContextV2.ContextLabelWeight` to handle new diagram generation labels.

## Example of diagram output

```
Class map of deferred tax
```

```mermaid
classDiagram
  direction LR

  class DeferredTax {
      +pricing_policy(product: Product, stockrecord: StockRecord) : Price
      +parent_pricing_policy(product: Product, children_stock: List~StockRecord~) : Price
  }
  class Product {
  }
  class StockRecord {
  }
  class Price {
  }

  class FixedPrice {
      +currency: str
      +excl_tax: Decimal
  }
  FixedPrice --> Price

  class UnavailablePrice {
  }
  UnavailablePrice --> Price

  DeferredTax --> Product
  DeferredTax --> StockRecord
  DeferredTax --> Price
```

### CODE REVIEW

**Comment by kgilpin** [comment]
Here's a list of code changes that are either not accounted for or related directly to the provided plan:

1. **Project Information Context Modification (ContextService)**:
    - **File**: `packages/navie/src/context.ts`
    - **Change**: Added the `GenerateDiagram` context label.

    ```typescript
    export enum ContextLabelName {
      HelpWithAppMap = 'help-with-appmap',
      GenerateDiagram = 'generate-diagram',
      // other labels...
    }
    ```

2. **Test Enhancements**:
    - **Files**:
        - `packages/navie/test/lib/mermaid-filter.spec.ts`
        - `packages/navie/test/lib/mermaid-fixer.spec.ts`
        - `packages/navie/test/lib/mermaid-validator.spec.ts`

    **Changes**: Added tests for `MermaidFilter`, `MermaidFixer`, and `MermaidValidator`.

3. **New Dependencies**:
    - **Files**:
        - `packages/navie/package.json`
    
    **Change**: Added the `mermaid` library as a new dependency.

    ```json
    "mermaid": "^9",
    ```

4. **Detailed Test Agent Selection**:
    - **Files**:
        - `packages/navie/test/services/agent-selection-service.spec.ts`
    
    **Change**: Added tests to ensure the correct agent selection, including the handling of multiple classifiers and preference for `ExplainAgent`.

5. **Refined `ExplainAgent` Prompt**:
    - **Files**:
        - `packages/navie/src/agents/explain-agent.ts`

    **Changes**: Updated prompt instructions in `ExplainAgent`.

    ```typescript
    const EXPLAIN_AGENT_PROMPT = `**Task: Answering User Questions about a Code Base**
    ```

6. **Service Adjustments**:
    - **Files**:
        - `packages/navie/src/navie.ts`
    
    **Change**: Minor enhancements and integration updates for services.

    ```typescript
    const agentSelectionService = new AgentSelectionService(
      interactionHistory,
      vectorTermsService,
      lookupContextService,
      applyContextService,
      techStackService,
      completionService
    );
    ```

7. **Completion Service Edits**:
    - **Files**:
        - `packages/navie/src/services/completion-service.ts`
    
    **Change**: Updated `Usage` type to export for external use.

    ```typescript
    export interface Usage extends Required<TokenUsage> {
      cost?: number;
    }
    ```

These changes either enhance existing functionality, ensure robustness through testing, refine prompts, or make small adjustments that might not be explicitly mentioned in the plan but are necessary for the new DiagramAgent feature to function correctly.

@evlawler this is for you :-)

File: packages/navie/src/agents/diagram-agent.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import { PROMPTS, PromptType } from '../prompt';
+import ContextService from '../services/context-service';
+import MermaidFixerService from '../services/mermaid-fixer-service';
+import { DIAGRAM_FORMAT_PROMPT } from './explain-agent';
+
+export const DIAGRAM_AGENT_PROMPT = `**Task: Generation of Software Diagrams**
+
+**About you**
+
+Your name is Navie. You are an AI softwrare architect created and maintained by AppMap Inc, and are available to AppMap users as a service.
```
Comment (dustinbyrne) [inline packages/navie/src/agents/diagram-agent.ts:15]: `softwrare`

File: packages/navie/src/lib/mermaid-filter.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    // When a new chunk is received, append the chunk to the buffer.
+    // Extract any complete lines from the buffer and process them.
+
+    this.buffer += chunk;
+
+    while (this.buffer.indexOf('\n') !== -1) {
+      const nextLine = this.buffer.slice(0, this.buffer.indexOf('\n'));
+      this.buffer = this.buffer.slice(this.buffer.indexOf('\n') + 1);
+      yield* this.processLine(nextLine);
+    }
```
Comment (dustinbyrne) [inline packages/navie/src/lib/mermaid-filter.ts:33]: `nextLine` is off by one, dropping the new line.

```suggestion
    for (;;) {
      const lineEndIndex = this.buffer.indexOf('\n');
      if (lineEndIndex === -1) break;

      const nextLine = this.buffer.slice(0, lineEndIndex + 1);
      this.buffer = this.buffer.slice(lineEndIndex + 1);
      yield* this.processLine(nextLine);
    }
```

Comment (kgilpin) [reply packages/navie/src/lib/mermaid-filter.ts]: It seems to be adding extra newlines: https://github.com/getappmap/appmap-js/actions/runs/9963827482/job/27530751071#step:5:2601

**Review by dustinbyrne — APPROVED** [review]

File: packages/navie/src/lib/mermaid-filter.ts (hunk 2 above)
Comment (kgilpin) [reply packages/navie/src/lib/mermaid-filter.ts:33]: @dustinbyrne is there still a code change required here or did it turn out to be ok as written?

**Pull request merged**